### ISSUE TITLE
Implementation of CSDI for time series forecasting

### ISSUE THREAD

**Issue by Super-Shen** [ermongroup/CSDI#4]
I am very interested in the research in your field and want to further explore the research. However, I can not get the same performance as the results on time series forecasting task described in the paper. Our result on electricity dataset for CRPS is 0.1594 while yours is 0.017. Could you please figure out the differences of experimental settings between mine and yours.

The implementation and experimental settings are as follows. First, we use the code from diff_models.py and main_model.py as the time series forecasting model, which is the same as your paper. As for the dataset settings, we use all of the 370 dimensions (clients) of electricity dataset and set prediction step as 24 as the settings in the related work[1]. We also applied test pattern strategy for target choice, which is written by ourselves. For each instance of time series data, we consider the first 24 steps as the observed part and the last 1 step as the target part to forecast the target value autoregressively.

I can not find the reason causing the difference between the performance. Could you please share the code of time series forecasting if it is available.

**Comment by y-tashi** [comment]
Hi, thank you for your interest in our work. Since our codes for forecasting are unstructured, I'd like to explain some key points for the implementation. I hope the following explanation helps you.
 
A main difference is the shape of instances. In our setting, the shape of each instance has (192, 370). The first 168 steps are the observed part and the last 24 step are the target part. Please see the following pseudocode:

> https://drive.google.com/file/d/1vy1Mez_NhRoQAsR_6OeOAT-lBfOsaoAm/view?usp=sharing

Since the instance shape is large, we used two techniques to reduce the usage of memory. One is subset sampling of features (see Appendix E.2). For each instance in training, we sample 64 features from 370 as follows:

>         sampled_data = []
>         sampled_mask = []
>         sampled_feature_id = []
>         sampled_gt_mask = []
>         size = 64
>         for i in range(len(observed_data)):
>             ind = np.arange(370)
>             np.random.shuffle(ind)
>             sampled_data.append(observed_data[i,...,ind[:size]])
>             sampled_mask.append(observed_mask[i,...,ind[:size]])
>             sampled_feature_id.append(feature_id[i,ind[:size]])
>             sampled_gt_mask.append(gt_mask[i,...,ind[:size]])
>         observed_data = torch.stack(sampled_data,0)
>         observed_mask = torch.stack(sampled_mask,0)
>         feature_id = torch.stack(sampled_feature_id,0)
>         gt_mask = torch.stack(sampled_gt_mask,0)

Here, "feature_id" is passed as an additional argument to get_side_info and used as follows:

>        feature_embed = self.embed_layer(feature_id)
>         feature_embed = feature_embed.unsqueeze(1).expand(-1,T,-1,-1)

We also adopted "linear attention transformer" package (see Appendix E.1) for get_torch_trans in diff_models.

**Comment by zhiyuanzhang15** [comment]
Hello, I am also very interested in utilizing CSDI for time series forecasting. However, I am currently facing difficulties with the dataset selection. I was wondering if you could provide assistance with this matter. Thank you very much.

**Comment by Capsar** [comment]
Hello, I also have a question regarding the datasets used for forecasting. It seems they are the same sets as used in TimeGrad & ScoreGrad (Also got the datasets from [Salinas et al.](https://github.com/mbohlkeschneider/gluon-ts/tree/mv_release/datasets)  and accessed through [GluonTS](https://github.com/awslabs/gluonts) However, the total time steps for Solar and Traffic seem different. Traffic total time steps are the same number of steps as Solar from TimeGrad.

TimeGrad:
![image](https://github.com/ermongroup/CSDI/assets/10345197/3658f23f-7112-4bfe-9d7d-f76185945f59)

CSDI:
![image](https://github.com/ermongroup/CSDI/assets/10345197/ddf64f92-6caa-4522-bd3a-8bd9ddcff512)

Is this a mistake? What are the exact descriptions for the datasets used for forecasting?

Best regards,
Caspar

**Comment by y-tashi** [comment]
Hi Capsar,
Yes, you are correct. The table 6 contains typos.
